templates/contribute/git.md

```diff
@@ -0,0 +1,274 @@
+
+Before starting, make sure you have:
+- Git installed on your computer
+- A GitHub account
+- (Optional but recommended) The GitHub CLI tool (`gh`) installed
+
+## Step 1: Fork the Repository on GitHub
+
+First, you need to create your own copy (fork) of the mathlib4 repository:
+
+1. Go to https://github.com/leanprover-community/mathlib4
+2. Click the "Fork" button in the top right corner
+3. Choose your GitHub account as the destination
+4. Wait for GitHub to create your fork
+
+You only ever need to do this step once. You can reuse your fork for many different branches and pull requests.
+
+## Step 2: Get a Local Copy of the Repository
+
+You have two options depending on whether you already have a clone of mathlib4:
+
+### Option A: If you don't have mathlib4 cloned yet
+
+Clone your fork (not the original repository):
+
+```bash
+# Replace YOUR_USERNAME with your GitHub username
+git clone https://github.com/YOUR_USERNAME/mathlib4.git
+cd mathlib4
+```
+
+This automatically sets up your fork as the `origin` remote, which is what we want.
+
+### Option B: If you already have mathlib4 cloned
+
+If you already have a clone from the original repository, you can reuse it. Just navigate to your existing mathlib4 directory:
+
+```bash
+cd path/to/your/existing/mathlib4
+```
+
+## Step 3: Set Up Remotes Correctly
+
+The remote setup depends on which option you chose above:
+
+### If you cloned your fork (Option A)
+
+You need to add the original repository as `upstream`:
+
+```bash
+# Add the original repository as 'upstream'
+git remote add upstream https://github.com/leanprover-community/mathlib4.git
+```
+
+### If you used an existing clone (Option B)
+
+You need to rename the existing remote and add your fork:
+
+```bash
+# Rename the existing remote to 'upstream'
+git remote rename origin upstream
+
+# Add your fork as 'origin' (replace YOUR_USERNAME with your GitHub username)
+git remote add origin https://github.com/YOUR_USERNAME/mathlib4.git
+```
+
+### Verify Your Remotes
+
+Regardless of which option you chose, verify your remotes are set correctly:
+
+```bash
+git remote -v
+```
+
+You should see:
+```
+origin    https://github.com/YOUR_USERNAME/mathlib4.git (fetch)
+origin    https://github.com/YOUR_USERNAME/mathlib4.git (push)
+upstream  https://github.com/leanprover-community/mathlib4.git (fetch)
+upstream  https://github.com/leanprover-community/mathlib4.git (push)
+```
+
+## Step 4: Configure the Master Branch
+
+Make sure your `master` branch tracks `upstream/master`:
+
+```bash
+git branch --set-upstream-to=upstream/master master
+```
+
+## Step 5: Keep Your Master Branch Up to Date
+
+Always keep your `master` branch synchronized with upstream:
+
+```bash
+git checkout master
+git pull
+```
+
+Run these commands regularly, especially before creating new branches.
+
+## Step 6: Prevent Accidental Commits to Master
+
+To avoid accidentally committing directly to `master`, you can set up a pre-commit hook:
+
+```bash
```

Comment:
All these command-line code snippets are marked as `bash`. I guess most are one-liners that would work on any shell, but this one is clearly an exception. I'm unsure if it would work on `zsh`, which is the default shell on Mac OS, and it almost certainly wouldn't work on the Windows command line.

Reply:
Just verified this works fine on zsh.

I think the best we can do here is suggest googling for advice on how to set up a git hook on Windows. It is probably not easy unless you have a unix-like shell, and the advice is going to be different for every GUI that allows it (if any even do).

@loefflerd, could you investigate this and suggest a set of instructions that work for you?

Reply:
Anyone who has installed Lean 4 following the official instructions has [Git for Windows](https://gitforwindows.org/) installed, which comes with Git Bash, which is, as the name suggests, a Unix-Like shell.

Reply:
Looks like we need a  line`chmod +x .git/hooks/pre-commit` or something.